scarlet-stream-adapter-coroutines/src/test/java/com/tinder/scarlet/streamadapter/coroutines/FlowStreamAdapterTest.kt

```diff
@@ -0,0 +1,98 @@
+package com.tinder.scarlet.streamadapter.coroutines
+
+import com.google.common.truth.Correspondence
+import com.google.common.truth.Truth.assertThat
+import com.tinder.scarlet.Stream
+import com.tinder.scarlet.WebSocket
+import com.tinder.scarlet.testutils.TestStreamObserver
+import com.tinder.scarlet.testutils.any
+import com.tinder.scarlet.testutils.containingBytes
+import com.tinder.scarlet.testutils.containingText
+import com.tinder.scarlet.testutils.test
+import com.tinder.scarlet.ws.Receive
+import com.tinder.scarlet.ws.Send
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.take
+import kotlinx.coroutines.flow.toList
+import kotlinx.coroutines.test.runTest
+import org.junit.Rule
+import org.junit.Test
+import java.util.Arrays
+
+class FlowStreamAdapterTest {
+
+    @get:Rule
+    val clientServerModel = object : ClientServerModel<Service>(Service::class.java) {
+        override fun observeWebSocketEvents(service: Service): TestStreamObserver<WebSocket.Event> {
+            return service.observeEvents().test()
+        }
+    }
+
+    @Test
+    fun `adapt - given a stream of strings, provides a Flow interface bound to the stream`() = runTest(dispatchTimeoutMs = 1000) {
+        // Given
+        val (client, server) = clientServerModel.givenConnectionIsEstablished()
+        val textMessage1 = "Hello"
+        val textMessage2 = "Hi!"
+        val bytesMessage1 = "Yo".toByteArray()
+        val bytesMessage2 = "Sup".toByteArray()
+        val testTextChannel = server.observeText()
+        val testBytesChannel = server.observeBytes()
+
+        // When
+        client.sendText(textMessage1)
+        val isSendTextSuccessful = client.sendTextAndConfirm(textMessage2)
+        client.sendBytes(bytesMessage1)
+        val isSendBytesSuccessful = client.sendBytesAndConfirm(bytesMessage2)
+
+        // Then
+        assertThat(isSendBytesSuccessful).isTrue()
+        assertThat(isSendTextSuccessful).isTrue()
+
+        clientServerModel.awaitValues(
+            any<WebSocket.Event.OnConnectionOpened<*>>(),
+            any<WebSocket.Event.OnMessageReceived>().containingText(textMessage1),
+            any<WebSocket.Event.OnMessageReceived>().containingText(textMessage2),
+            any<WebSocket.Event.OnMessageReceived>().containingBytes(bytesMessage1),
+            any<WebSocket.Event.OnMessageReceived>().containingBytes(bytesMessage2)
+        )
+
+        assertThat(testTextChannel.take(2).toList()).containsExactly(
```

Comment:
I do not fully follow the tests, but is this `testTextChannel` caching values that it may have already received?

Reply:
so the testTextChannel, (renamed to flow at this point) hasn't been read yet however since the connection is still open during this evaluation, we only want the first 2 values otherwise we would block indefinitely.